src/game/logic/map/polygontrigger.cpp

```diff
@@ -21,28 +24,417 @@
 PolygonTrigger *PolygonTrigger::s_thePolygonTriggerListPtr;
 #endif
 
-bool PolygonTrigger::Parse_Polygon_Triggers_Data_Chunk(DataChunkInput &file, DataChunkInfo *info, void *userdata)
+int PolygonTrigger::s_currentID;
+
+PolygonTrigger::PolygonTrigger(int initial_allocation) :
+    m_nextPolygonTrigger(nullptr),
+    m_points(nullptr),
+    m_numPoints(0),
+    m_sizePoints(0),
+    m_riverStart(0),
+    m_exportWithScripts(false),
+    m_isWaterArea(false),
+    m_isRiver(false),
+    m_isShownInLayer(true),
+    m_isSelected(false)
 {
-#ifdef GAME_DLL
-    return Call_Function<bool, DataChunkInput &, DataChunkInfo *, void *>(
-        PICK_ADDRESS(0x00570D60, 0x006CE6F6), file, info, userdata);
-#else
-    return 0;
-#endif
+    if (initial_allocation < 2) {
+        initial_allocation = 2;
+    }
+
+    m_points = new ICoord3D[initial_allocation];
+    m_sizePoints = initial_allocation;
+    m_triggerID = s_currentID++;
+    m_waterHandle.m_polygon = this;
+}
+
+PolygonTrigger::~PolygonTrigger()
+{
+    if (m_points) {
+        delete[] m_points;
+        m_points = nullptr;
+    }
+
+    if (m_nextPolygonTrigger) {
+        PolygonTrigger *next;
+
+        for (PolygonTrigger *p = m_nextPolygonTrigger; p != nullptr; p = next) {
+            next = p->Get_Next();
+            p->Set_Next(nullptr);
+            p->Delete_Instance();
+        }
+    }
+}
+
+void PolygonTrigger::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char version = 1;
+    xfer->xferVersion(&version, 1);
+    xfer->xferInt(&m_numPoints);
+
+    for (int i = 0; i < m_numPoints; i++) {
+        xfer->xferICoord3D(&m_points[i]);
+    }
+
+    xfer->xferIRegion2D(&m_bounds);
+    xfer->xferReal(&m_radius);
+    xfer->xferBool(&m_boundsNeedsUpdate);
+}
+
+void PolygonTrigger::Reallocate()
+{
+    captainslog_dbgassert(m_numPoints <= m_sizePoints, "Invalid m_numPoints.");
+
+    if (m_numPoints == m_sizePoints) {
+        m_sizePoints *= 2;
+        ICoord3D *points = new ICoord3D[m_sizePoints];
+
+        for (int i = 0; i < m_numPoints; i++) {
+            points[i] = m_points[i];
+        }
+
+        delete[] m_points;
+        m_points = points;
+    }
+}
+
+void PolygonTrigger::Update_Bounds() const
+{
+    m_bounds.lo.y = 0x7FFFF0;
+    m_bounds.lo.x = 0x7FFFF0;
+    m_bounds.hi.y = 0xFF800010;
+    m_bounds.hi.x = 0xFF800010;
+
+    for (int i = 0; i < m_numPoints; i++) {
+        if (m_points[i].x < m_bounds.lo.x) {
+            m_bounds.lo.x = m_points[i].x;
+        }
+
+        if (m_points[i].y < m_bounds.lo.y) {
+            m_bounds.lo.y = m_points[i].y;
+        }
+
+        if (m_points[i].x > m_bounds.hi.x) {
+            m_bounds.hi.x = m_points[i].x;
+        }
+
+        if (m_points[i].y > m_bounds.hi.y) {
+            m_bounds.hi.y = m_points[i].y;
+        }
+    }
+
+    m_boundsNeedsUpdate = false;
+    float x = (float)(m_bounds.hi.x - m_bounds.lo.x) / 2.0;
+    float y = (float)(m_bounds.lo.y + m_bounds.hi.y) / 2.0;

```

Comment:
Does compiler make it a division op? If yes, then `* 0.5f` would be cheaper.

Reply:
It should optimize to multiplication, i don't think i've seen a single case where it doesn't(when its not Od)